src/lib/journald.ts

```diff
@@ -0,0 +1,138 @@
+import { ChildProcess, spawn } from 'child_process';
+import { EventEmitter } from 'events';
+import { isArray } from 'lodash';
+import StrictEventEmitter from 'strict-event-emitter-types';
+
+import log from './supervisor-console';
+
+const CHROOT_COMMAND = 'chroot';
+const CHROOT_DIRECTORY = '/mnt/root';
+const JOURNALD_COMMAND = 'journalctl';
+const JOURNALD_FORMAT = 'json';
+const JOURNALD_FORMAT_FLAG = '-o';
+const JOURNALD_FOLLOW_FLAG = '-f';
+const JOURNALD_SERVICE_NAME_FLAG = '-u';
+const JOURNALD_COUNT_FLAG = '-n';
+
+export interface Events {
+	logLine: {
+		message: string;
+		priority: number;
+		name: string;
+		timestamp: Date;
+
+		// We also forward all other fields which we get
+		[key: string]: unknown;
+	};
+	stderrLine: string;
+	disconnect: { code: number; signal: string };
+	error: Error;
+}
+
+export type JournaldEventEmitter = StrictEventEmitter<EventEmitter, Events>;
+
+export class Journald extends (EventEmitter as new () => JournaldEventEmitter) {
+	private partialLog: string = '';
+
+	public constructor(private process: ChildProcess) {
+		super();
+	}
+
+	public end() {
+		this.process.kill();
+	}
+
+	public addData(data: string) {
+		// split on newlines, storing any data which is not
+		// followed by a newline
+		data = this.partialLog.concat(data);
+		this.partialLog = '';
+		let index = data.indexOf('\n');
+		while (index !== -1) {
+			const message = data.slice(0, index);
+			this.emitMessage(message);
+
+			data = data.slice(index + 1);
+			index = data.indexOf('\n');
+		}
+
+		this.partialLog = data;
+	}
+
+	private emitMessage(message: string) {
+		try {
+			const jsonMsg = JSON.parse(message);
+			this.emit('logLine', {
+				message: Journald.processMessageField(jsonMsg.MESSAGE),
+				priority: parseInt(jsonMsg.PRIORITY, 10),
+				// __REALTIME_TIMESTAMP is in microseconds, and JS
+				// expects millis
+				timestamp: new Date(parseInt(jsonMsg.__REALTIME_TIMESTAMP, 10) / 1000),
+				// We try our best to get a decent name for a log
+				name:
+					jsonMsg.SYSLOG_IDENTIFIER ||
+					jsonMsg._SYSTEMD_UNIT ||
+					jsonMsg.CONTAINER_NAME,
+
+				...jsonMsg,
+			});
+		} catch (e) {
+			this.emit('error', e);
+		}
+	}
+
+	private static processMessageField(message: string | number[] | null) {
+		if (message == null) {
+			return 'Message too large, run with { "all": true }';
+		} else if (isArray(message)) {
+			return String.fromCharCode(...message);
+		} else {
+			return message;
+		}
+	}
+}
+
+export function getJournald(opts: {
+	follow: boolean;
+	count?: number;
+	unit?: string;
+}): Journald {
+	const args = [
+		CHROOT_DIRECTORY,
+		JOURNALD_COMMAND,
+		JOURNALD_FORMAT_FLAG,
+		JOURNALD_FORMAT,
```

Comment:
NIT:
```typescript
const args = ['/mnt/root', 'journalctl', '--output', 'json']
```
would require less effort to read. The constants are only used in one place.

Reply:
Yeah fair point, I had thought I'd need to do something a bit more complex, but as it turns out I didn't.